matchbox/provider_test.go

```diff
@@ -0,0 +1,251 @@
+package matchbox
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"net"
+	"testing"
+
+	"github.com/coreos/matchbox/matchbox/rpc"
+	"github.com/coreos/matchbox/matchbox/server"
+	"github.com/coreos/matchbox/matchbox/storage"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/hashicorp/terraform/terraform"
+)
+
+var providers = map[string]terraform.ResourceProvider{
+	"matchbox": Provider(),
+}
+
+func TestProvider(t *testing.T) {
+	if err := Provider().(*schema.Provider).InternalValidate(); err != nil {
+		t.Fatalf("err: %s", err)
+	}
+}
+
+type MatchboxFixtureServer struct {
+	Store    storage.Store
+	Listener net.Listener
+}
+
+func NewMatchboxFixtureServer(s storage.Store) *MatchboxFixtureServer {
+	l, err := net.Listen("tcp", "127.0.0.1:0")
+	if err != nil {
+		panic(fmt.Errorf("failed to start listener: %v", err))
+	}
+
+	return &MatchboxFixtureServer{
+		Store:    s,
+		Listener: l,
+	}
+}
+
+func (s *MatchboxFixtureServer) Start() {
+	cfg, err := testServerTLSConfig()
+	if err != nil {
+		panic(fmt.Errorf("Invalid TLS credentials: %v", err))
+	}
+
+	srv := server.NewServer(&server.Config{s.Store})
+	rpc := rpc.NewServer(srv, cfg)
+	go rpc.Serve(s.Listener)
+}
+
+func (s *MatchboxFixtureServer) AddProviderConfig(hcl string) string {
+	return fmt.Sprintf(`
+		provider "matchbox" {
+			endpoint = "%s"
+			client_cert = <<CERT
+%s
+CERT
+			client_key = <<KEY
+%s
+KEY
+			ca         = <<CA
+%s
+CA
+		}
+
+		%s
+	`, s.Listener.Addr().String(),
+		fixtureClientCertPEMBlock,
+		fixtureClientKeyPEMBlock,
+		fixtureCAPEMBlock,
+		hcl,
+	)
+}
+
+func testServerTLSConfig() (*tls.Config, error) {
```

Comment:
Shoot you're right. That should be made nicer in the main repo. For now, its fine to mirror the exact logic used upstream, but accepting raw byte slices instead of filenames.